core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/ruin/RuinPhaseCommand.java

```diff
@@ -0,0 +1,65 @@
+package org.optaplanner.core.impl.ruin;
+
+import org.optaplanner.core.api.score.director.ScoreDirector;
+import org.optaplanner.core.impl.heuristic.move.Move;
+import org.optaplanner.core.impl.heuristic.move.NoChangeMove;
+import org.optaplanner.core.impl.heuristic.selector.move.MoveSelector;
+import org.optaplanner.core.impl.phase.custom.CustomPhaseCommand;
+import org.optaplanner.core.impl.phase.event.PhaseLifecycleListener;
+import org.optaplanner.core.impl.phase.scope.AbstractPhaseScope;
+import org.optaplanner.core.impl.phase.scope.AbstractStepScope;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+
+/**
+ * Ruins solution.
+ */
+public class RuinPhaseCommand<Solution_> implements CustomPhaseCommand<Solution_>, PhaseLifecycleListener<Solution_> {
+
+    private final MoveSelector<Solution_> moveSelector;
+
+    public RuinPhaseCommand(MoveSelector<Solution_> moveSelector) {
+        this.moveSelector = moveSelector;
+    }
+
+    @Override
+    public void changeWorkingSolution(ScoreDirector<Solution_> scoreDirector) {
+        Move<Solution_> nextRuinMove = moveSelector.iterator().next();
+        if (nextRuinMove != null) {
+            nextRuinMove.doMoveOnly(scoreDirector);
+            scoreDirector.triggerVariableListeners();
+        } else {
+            NoChangeMove<Solution_> noChangeMove = new NoChangeMove<>();
+            noChangeMove.doMoveOnly(scoreDirector);
```

Comment:
Should this even be possible? A ruin phase that does nothing?
Shouldn't we rather throw an exception here?

Reply:
when you run ruin phase 5 times in a row with lets say 20% of ruin size and then you try to apply ruin once more then the iterator will search for entities that are not yet ruined. Since 5*20%~100% is already ruined then it will find no more entities left to be ruined. In this case the iterator returns null and no action is taken.

Reply:
I left a comment as well